### PR TITLE
[quant] Run weight_post_process for QAT

### PR DESCRIPTION
**Pull request by supriyar**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #34232 [quant] Fix histogram observer to work with QAT on GPU
* **#33852 [quant] Run weight_post_process for QAT**

Summary:
This fixes an issue for QAT models. During eval if we call `prepare_qat` and `convert` before calling `load_state_dict` it throws an error because the weight info (num channels) is not updated in the observer module.
It is not an issue for per-tensor case

Fixes issue #33830

Test Plan:
python test/test_quantization.py EagerModePostTrainingQuantTest.test_eval_after_train
python test/test_quantization.py EagerModeQuantizationAwareTrainingTest.test_eval_after_train

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20212996](https://our.internmc.facebook.com/intern/diff/D20212996)

### CODE REVIEW

File: torch/nn/quantized/modules/conv.py [hunk 1]
```diff
@@ -232,6 +232,7 @@ def from_float(cls, mod):
             assert hasattr(mod, 'activation_post_process'), \
                 'Input QAT module must have observer attached'
             weight_post_process = mod.weight_fake_quant
+            weight_post_process(mod.weight)
```
Comment (jerryzh168) [inline torch/nn/quantized/modules/conv.py]: can you just move this outside of if?

File: torch/nn/quantized/modules/linear.py [hunk 2]
```diff
@@ -204,6 +204,7 @@ def from_float(cls, mod):
         if hasattr(mod, 'weight_fake_quant'):
             # assert type(mod) == QATLinear, 'training mode nnq.Linear.from_float only works for nn.qat.Linear'
             weight_post_process = mod.weight_fake_quant
+            weight_post_process(mod.weight)
```
Comment (jerryzh168) [inline torch/nn/quantized/modules/linear.py]: similar for this one

File: test/test_quantization.py [hunk 3]
```diff
@@ -793,6 +813,31 @@ def checkQuantized(model):
         model = quantize_qat(model, test_only_train_fn, self.img_data)
         checkQuantized(model)
 
+    def test_eval_after_train(self):
```
Comment (jerryzh168) [inline test/test_quantization.py]: I think maybe include all steps would be better: e.g. test_train_save_load_eval

File: test/test_quantization.py [hunk 4]
```diff
@@ -793,6 +813,31 @@ def checkQuantized(model):
...
 
+    def test_train_save_load_eval(self):
+        qengine = 'fbgemm'
+        model = ManualLinearQATModel()
+        model = prepare_qat(model)
+        self.checkObservers(model)
+        test_only_train_fn(model, self.train_data)
+        model = convert(model)
+
+        model = quantize_qat(ManualLinearQATModel(), test_only_train_fn,
```
Comment (raghuramank100) [inline test/test_quantization.py]: Looks like lines 818-822 are redundant as you are creating a model afresh here.

File: test/test_quantization.py [hunk 5]
```diff
@@ -793,6 +813,31 @@ def checkQuantized(model):
         model = quantize_qat(model, test_only_train_fn, self.img_data)
         checkQuantized(model)
 
+    def test_train_save_load_eval(self):
+        qengine = 'fbgemm'
```
Comment (raghuramank100) [inline test/test_quantization.py]: Can you sweep over both qengines here?

File: test/test_quantization.py [hunk 6]
```diff
@@ -793,6 +813,31 @@ def checkQuantized(model):
         model = quantize_qat(model, test_only_train_fn, self.img_data)
         checkQuantized(model)
 
+    def test_train_save_load_eval(self):
```
Comment (raghuramank100) [inline test/test_quantization.py]: Should this function be called test_train_save_load_qat(self):?

File: test/test_quantization.py [hunk 7]
```diff
@@ -311,6 +311,26 @@ def checkQuantized(model):
 
         checkQuantized(model)
 
+    def test_train_save_load_eval(self):
+        model = AnnotatedTwoLayerLinearModel()
```
Comment (raghuramank100) [inline test/test_quantization.py]: Suggest sweeping over multiple qconfigs here (cover both per tensor and per channel quant)

File: test/test_quantization.py [hunk 8]
```diff
@@ -311,6 +311,26 @@ def checkQuantized(model):
 
         checkQuantized(model)
 
+    def test_train_save_load_eval(self):
```
Comment (raghuramank100) [inline test/test_quantization.py]: Should this be called test_post_train_quant_save_load(self):?

File: test/test_quantization.py (hunk 6 above)
Comment (supriyar) [reply test/test_quantization.py]: The class has information about PTQ vs QAT  - `EagerModeQuantizationAwareTrainingTest`

File: test/test_quantization.py (hunk 8 above)
Comment (supriyar) [reply test/test_quantization.py]: The class has information about PTQ vs QAT - `EagerModePostTrainingQuantTest`

File: test/test_quantization.py [hunk 9]
```diff
@@ -793,6 +813,31 @@ def checkQuantized(model):
...
+        # Check eval using quantized state_dict
+        model = ManualLinearQATModel()
+
+        model.qconfig = torch.quantization.get_default_qat_qconfig(qengine)
+        torch.quantization.prepare_qat(model, inplace=True)
+        torch.quantization.convert(model, inplace=True)
+        model.load_state_dict(quant_state_dict)
+
+        out = model(x)
+        self.assertEqual(ref, out)
```
Comment (raghuramank100) [inline test/test_quantization.py]: How about testing for consistency at multiple points: 
1. Consistency of fake-quant state dict with loading a model from scratch
2. Consistency of state dict after convert with a model created using PTQ (the qconfig needs to be consistent: Both FQ and observer need to use the same type of quantization).

**Review by raghuramank100 — CHANGES_REQUESTED** [review]
Good find! I have a couple of suggestions to improve the tests. Thanks

File: test/test_quantization.py [hunk 10]
```diff
@@ -311,6 +311,41 @@ def checkQuantized(model):
 
         checkQuantized(model)
 
+    @given(qengine=st.sampled_from(("qnnpack", "fbgemm")))
```
Comment (raghuramank100) [inline test/test_quantization.py]: NIT: Can we have a better name than train_save_load_eval? There is no training here.

File: test/test_quantization.py [hunk 11]
```diff
@@ -793,6 +828,43 @@ def checkQuantized(model):
...
+        with override_quantized_engine(qengine):
+            model = TwoLayerLinearModel()
+            model = torch.quantization.QuantWrapper(model)
+            model.qconfig = torch.quantization.get_default_qat_qconfig(qengine)
+            model = prepare_qat(model)
+
+            fq_state_dict = model.state_dict()
+
+            test_only_train_fn(model, self.train_data)
+            model = convert(model)
```
Comment (raghuramank100) [inline test/test_quantization.py]: Dont you need to call model.eval() here?

File: test/test_quantization.py [hunk 12]
```diff
@@ -793,6 +828,43 @@ def checkQuantized(model):
...
+            model = TwoLayerLinearModel()
+            model = torch.quantization.QuantWrapper(model)
+            model.qconfig = torch.quantization.get_default_qat_qconfig(qengine)
+            torch.quantization.prepare_qat(model, inplace=True)
+            new_state_dict = model.state_dict()
+
+            # Check to make sure the model after prepare_qat has the same state_dict as original.
+            self.assertEqual(set(fq_state_dict.keys()), set(new_state_dict.keys()))
+
+            torch.quantization.convert(model, inplace=True)
```
Comment (raghuramank100) [inline test/test_quantization.py]: Similar comment: you need to call model.eval() here

File: test/test_quantization.py [hunk 13]
```diff
@@ -793,6 +828,43 @@ def checkQuantized(model):
...
+            model.load_state_dict(quant_state_dict)
```
Comment (raghuramank100) [inline test/test_quantization.py]: How about adding one more check here?
```python
            model = TwoLayerLinearModel()
            model.eval()
            model = torch.quantization.QuantWrapper(model)
            model.qconfig = torch.quantization.get_default_qconfig(qengine)
            torch.quantization.prepare(model, inplace=True)
            torch.quantization.convert(model, inplace=True)
            self.assertEqual(set(model.state_dict().keys()), set(quant_state_dict.keys()))
            out = model(x)
            self.assertEqual(ref, out)
```

**Review by raghuramank100 — CHANGES_REQUESTED** [review]
Looks good, adding a request for additional testing.

**Review by raghuramank100 — APPROVED** [review]
Looks good, thanks for making the changes!

**Pull request closed without merging**